code/modules/mob/living/silicon/ai/ai_programs.dm

```diff
@@ -0,0 +1,1017 @@
+		var/obj/mecha/T = target
+		T.cell.give(power_sent)
+	if(istype(target, /mob/living/silicon/robot))
+		var/mob/living/silicon/robot/T = target
+		T.cell.give(power_sent)
+	if(istype(target, /obj/machinery/power/apc))
+		var/obj/machinery/power/apc/T = target
+		T.cell.give(power_sent)
+	if(istype(target, /mob/living/carbon/human/machine))
+		var/mob/living/carbon/human/machine/T = target
+		T.adjust_nutrition(AI.adapter_efficiency * (power_sent / 10))
+
+	// Handles draining the SMES
+	for(var/obj/machinery/power/smes/power_source in A)
+		if(power_source.charge < power_sent)
+			continue
+		power_source.charge -= power_sent
+		break
+	AI.program_picker.nanites -= 20
+	AI.play_sound_remote(target, 'sound/goonstation/misc/fuse.ogg', 50)
+	camera_beam(target, "rped_upgrade", 'icons/effects/effects.dmi', 5)
+
+/datum/spell/ai_spell/ranged/power_shunt/on_purchase_upgrade()
+	power_sent = min(10000, 2500 + (spell_level * 2500))
+	cooldown_handler.recharge_duration = max(min(base_cooldown, base_cooldown - (max(spell_level - 3, 0) * 60)), cooldown_min)
+
+// Repair Nanites - Uses large numbers of nanites to repair things
+/datum/ai_program/repair_nanites
+	program_name = "Repair Nanites"
+	program_id = "repair_nanites"
+	description = "Repair an APC, Borg, or Mech with large numbers of robotic nanomachines!"
+	cost = 2
+	nanite_cost = 75
+	power_type = /datum/spell/ai_spell/ranged/repair_nanites
+	unlock_text = "Repair nanomachine firmware installation complete!"
+
+/datum/spell/ai_spell/ranged/repair_nanites
+	name = "Repair Nanites"
+	desc = "Repair an APC, Borg, or Mech with large numbers of robotic nanomachines!"
+	action_icon = 'icons/obj/surgery.dmi'
+	action_icon_state = "tube"
+	ranged_mousepointer = 'icons/mecha/mecha_mouse.dmi'
+	auto_use_uses = FALSE
+	base_cooldown = 300 SECONDS
+	cooldown_min = 30 SECONDS
+	level_max = 10
+	selection_activated_message = "<span class='notice'>You prepare to order your nanomachines to repair...</span>"
+	selection_deactivated_message = "<span class='notice'>You rescind the order.</span>"
+
+/datum/spell/ai_spell/ranged/repair_nanites/cast(list/targets, mob/user)
+	var/target = targets[1]
+	if(!check_camera_vision(user, target))
+		revert_cast()
+		return
+	if(!istype(target, /mob/living/silicon/robot) && !istype(target, /obj/machinery/power/apc) && !istype(target, /obj/mecha) && !istype(target, /mob/living/carbon/human/machine))
+		to_chat(user, "<span class='warning'>You can only recharge borgs, mechs, and APCs!</span>")
+		revert_cast()
+		return
+	var/mob/living/silicon/ai/AI = user
+	if(istype(target, /mob/living/carbon/human/machine)  && !AI.universal_adapter)
+		to_chat(user, "<span class='warning'>This software lacks the required upgrade to recharge IPCs!</span>")
+		revert_cast()
+		return
+	if(istype(target, /obj/mecha)|| istype(target, /obj/machinery/power/apc))
+		var/obj/T = target
+		T.obj_integrity += min(T.max_integrity, T.max_integrity * (0.2 +  min(0.3, (0.1 * spell_level))))
```

Comment:
```suggestion
		T.obj_integrity += min(T.max_integrity, T.max_integrity * (0.2 + min(0.3, (0.1 * spell_level))))
```
could you also add a comment explaining the values here and generally what the calculation is meant to be doing? e.g. `// increase object integrity by 30%-50% depending on spell upgrade` (or whatever it actually is if i misread it), that way if someone goes back to look at it and thinks there's a logic error, they can see what your intent is versus what the code is doing in order to determine if there's a discrepancy